Fix panel selection after deleting or hiding beatmaps

nobbele opened the pull request:
Closes #33536 (including my comment)
Closes https://github.com/ppy/osu/issues/33512

Adds a test for deleting a set and hiding a difficult, making sure it selects the new panel that replaces it. 

The cleanest solution I found this was to implement it in two layers.
- Carousel will, when an entry is deleted, select the new entry that occupies the slot the old one was in. (I don't know if this is okay to place in the general Carousel container rather than BeatmapCarousel).
- SongSelect will, when an entry is invalidated, look for the next best difficulty in the same set, if there are valid difficulties in the set. Otherwise it will randomize the beatmap, although this shouldn't happen if the Carousel implementation works correctly.

The `findNextValidBeatmap` is a bit complicated, there might be a simpler way to express it, but basically it gets the next valid difficulty located after the newly hidden one. And one isn't found, it will find the previous valid difficulty before the newly one. Basically, finds the the difficulty closest to the newly hidden one.

Dealing with indexes directly is kinda ugly but it's the only thing I could come up with.

nobbele committed:
Fixes panel selection after deleting or hiding beatmaps

wip

nobbele commented:
The CodeFactor check wasn't from my changes, I'm not sure what the right `carouselItems?.Any() != true` equivalent is, `carouselItems == null || carouselItems.Count == 0`?

peppy commented:
Thanks for your contribution.

That said, I would highly recommend asking before submitting relatively complex fixes for features recently released by core team members. On principle, if this is not reviewable within 10 minutes I will be closing it and implement myself as this is already on my task list.

peppy commented:
> The CodeFactor failure wasn't from my changes, I'm not sure what the right `carouselItems?.Any() != true` equivalent is, `carouselItems == null || carouselItems.Count == 0`?

codefactor can generally be ignored, yeah.

peppy committed:
Merge branch 'master' into ssv2-33536

peppy committed:
Remove unnecessary `IndexOf` method

peppy committed:
Simplify index selection implementation

Reduces the total iterations of `Items` and better guards against edge
cases (like all items being deleted).

peppy commented on osu.Game/Screens/SelectV2/BeatmapCarousel.cs:
```diff
@@ -143,12 +143,28 @@ private void beatmapSetsChanged(object? beatmaps, NotifyCollectionChangedEventAr
                 case NotifyCollectionChangedAction.Remove:
-                    foreach (var set in oldItems!)
+
+                    int? selectionIndex = null;
```
@bdach if possible, could i get a quick smell test on this new logic? i want to get this in ASAP and since i rewrote this portion, would appreciate a second set of eyes to make sure i haven't done anything silly.

bdach replied:
The logic looks sound logically but in actual testing... it doesn't do what I think it should be doing? My answer to the TODO down there "should this exist in song select instead of here?" is yes because this implementation appears to be completely ignoring the sort mode? It's selecting the closest index item *as specified by the order in `Items` which is not *the actual active sorting order*.

peppy commented:
Closing this as the implementation was wrong in too many ways. New PR will arrive soon.
